File: tests/crunchy_vs_smooth.ts
import * as anchor from "@project-serum/anchor";
import { web3, Program } from "@project-serum/anchor";
import { assert } from "chai";
import { CrunchyVsSmooth } from "../target/types/crunchy_vs_smooth";

describe("crunchy_vs_smooth", () => {
  // Configure the client to use the local cluster.
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const program = anchor.workspace.CrunchyVsSmooth as Program<CrunchyVsSmooth>;

  const [voter, bump] = web3.PublicKey.findProgramAddressSync(
    [Buffer.from("Voter")],
    program.programId
  );

  console.log(bump);

  it("Initialized the voter!", async () => {
    await program.methods
      .init(bump)
      .accounts({
        owner: provider.wallet.publicKey,
        voter: voter,
        systemProgram: anchor.web3.SystemProgram.programId,
      })
      .rpc();

    let currentVoteAccountState = await program.account.voteAccount.fetch(
      voter
    );
    assert.equal(0, currentVoteAccountState.crunchy.toNumber());
    assert.equal(0, currentVoteAccountState.smooth.toNumber());
  });

  it("Vote Smooth", async () => {
    const voteSmooth = await program.methods
      .voteSmooth()
      .accounts({ vote: voter })
      .rpc();

    let currentVoteAccountState = await program.account.voteAccount.fetch(
      voter
    );

    assert.equal(1, currentVoteAccountState.smooth.toNumber());
  });

  it("Vote Crunchy", async () => {
    const voteCrunchy = await program.methods
      .voteCrunchy()
      .accounts({ vote: voter })
      .rpc();

    let currentVoteAccountState = await program.account.voteAccount.fetch(
      voter
    );

    assert.equal(1, currentVoteAccountState.crunchy.toNumber());
  });
});

File: programs_py/crunchy_vs_smooth.py
# crunchy_vs_smooth
# Built with Seahorse v0.1.6


from seahorse.prelude import *

# This will be updated once the project is build, your program id can be found at crunchy_vs_smooth/target/idl/[your_project_name].json
declare_id("Fo9d34XdUczXdNt9jkqrQseyUQys9bmTDC31utY3zt5x")


# Here we define all our instructions, each of the method below as an RPC end point which can be invoked by clients.
@instruction
def init(owner: Signer, voter: Empty[VoteAccount], vote_account_bump: u8):
    # As a new user connects, we create a new voter PDA account for him and intialize the account.
    init_voter = voter.init(payer=owner, seeds=["Voter", owner])
    # Assign the owner or the Signer of the one initialize the accouunt to the user's newly created VoteAccount owner.
    init_voter.owner = owner.key()
    # Assign the bump to the one initializing the accouunt to the user's newly created VoteAccount bump.
    init_voter.bump = vote_account_bump


# To vote crunchy
@instruction
def vote_crunchy(owner: Signer, vote: VoteAccount):
    # Check if the public key of the signer is the same as the owner in the vote account.
    assert owner.key() == vote.owner, "This is not your Vote account!"
    # Increment the crunchy variable in the user's VoteAccount
    vote.crunchy += 1


# To vote smooth
@instruction
def vote_smooth(owner: Signer, vote: VoteAccount):
    # Check if the public key of the signer is the same as the owner in the vote account.
    assert owner.key() == vote.owner, "This is not your Vote account!"
    # Increment the smooth variable in the user's VoteAccount
    vote.smooth += 1


# Defining the account which will be stored on-chain for every unique wallet interacting with our program.
class VoteAccount(Account):
    owner: Pubkey
    crunchy: u64
    smooth: u64
    bump: u8
